solarforecastarbiter/plotting/timeseries.py

```diff
@@ -0,0 +1,252 @@
+
+
+def make_quality_bars(flags, plot_width, x_range, source=None):
+    """Make the quality bar figures for observation validation"""
+    if source is None:
+        source = ColumnDataSource(flags)
+    palette = iter(PALETTE * 3)
+    out = []
+    for flag, color in FLAG_COLORS.items():
+        if flag not in flags or flags[flag].dropna().empty:
+            continue
+        if color is None:
+            color = next(palette)
+        nextfig = _single_quality_bar(flag, plot_width, x_range,
+                                      color, source)
+        out.append(nextfig)
+
+    # add the title to the top bar if there are any bars
+    if out:
+        out[0].plot_height = 60
+        out[0].title.text = 'Quality Flags'
+        out[0].title.text_font_size = '1em'
+
+    return out
+
+
+def add_hover_tool(fig, source, **hover_kwargs):
+    """Add a hover tool to fig. If `add_line=True` in `hover_kwargs`
+    an invisible line is added to enable hover values step plots"""
+    if hover_kwargs.pop('add_line', False):
+        fig.line(x='timestamp', y='value', source=source, line_alpha=0)
+
+    tooltips = [
+        ('timestamp', '@timestamp{%FT%H:%M:%S%z}'),
+        ('value', '@value{%0.2f}')
+    ]
+    if 'active_flags' in source.data.keys():
+        tooltips.append(('quality flags', '@active_flags'))
+    hover = HoverTool(tooltips=tooltips, formatters={
+        'timestamp': 'datetime', 'value': 'printf'}, mode='vline',
+                      **hover_kwargs)
+    fig.add_tools(hover)
+
+
+def make_basic_timeseries(values, object_name, variable, interval_label,
+                          plot_width, source=None):
+    """Make a basic timeseries plot (with either a step or line)
+    and add a hover tool"""
+    if source is None:
+        source = ColumnDataSource(values)
+    plot_method, plot_kwargs, hover_kwargs = plot_utils.line_or_step(
+        interval_label)
+    figure_title = build_figure_title(object_name, values.index[0],
+                                      values.index[-1])
+    fig = figure(title=figure_title, sizing_mode='scale_width',
+                 plot_width=plot_width,
+                 plot_height=300, x_range=(values.index[0], values.index[-1]),
+                 x_axis_type='datetime',
+                 tools='pan,wheel_zoom,box_zoom,zoom_in,zoom_out,reset,save',
+                 toolbar_location='above',
+                 min_border_bottom=50)
+    getattr(fig, plot_method)(x='timestamp', y='value', source=source,
+                              **plot_kwargs)
+    fig.yaxis.axis_label = plot_utils.format_variable_name(variable)
+    fig.xaxis.axis_label = 'Time (UTC)'
+    add_hover_tool(fig, source, **hover_kwargs)
+    return fig
+
+
+def _make_layout(figs):
+    layout = gridplot(figs,
+                      ncols=1,
+                      merge_tools=True,
+                      toolbar_location='above',
+                      sizing_mode='scale_width')
+    return components(layout)
```

Comment:
Why not let the dashboard call `components`? Or make a `to_components` decorator that takes a bool kwarg for the functions below. My understanding is that calling components here prevents reuse for things like bokeh apps or creating static files.